docs/design/storagenode-graceful-exit.md

```diff
@@ -91,10 +92,169 @@ This process including the Storage nodes transferring their pieces to other node
+	  - Join returns after one full metainfo loop, or on err. On success, the service should increment `nodes.exit_loop_count` for the nodeIDs that were processed in the full loop.
+    - Execution intervals and requiredMetainfoLoops should be configurable
+  - Add CheckStatus loop
+    - Queries `exit_pieceinfo` grouping by node ID and counts completed, and incomplete.  If all records are complete and `nodes.exit_loop_count` <= `requiredMetainfoLoops`
+      - Set `nodes.exit_completed` to current time.
+      - Remove all `exit_pieceinfos` for the completed node ID
+    - Execution intervals should be configurable
+- Create GracefulExit endpoint
+  - Endpoints should be secured using the peer Identity provided in context
+  - InitiateExit initiates the exit process by setting `nodes.exit_initiated_at` to current time
+  - GetPutOrders
+    - Queries `nodes` by `id` and `exit_completed`.
+      - If completed it returns `PutOrdersResponse` with `exit_completed` set to true and `exit_completed_signature` set to a satellite signed value of satellite id, storage, node id, "gracefuleexit-successfully-complete", completed timestamp.
+      - Else it checks whether there are any records in `exit_pieceinfo` for the node ID. If not, it returns a `PutOrdersResponse` with `next_execution` set to now + 1hr (this should be configurable, and in the future may be calculated). This covers the case where the metainfo loop hasn't added records into `exit_pieceinfo` yet. // TODO: maybe we return a time when the node should check again.
+      - Else it checks whether there are any records in `exit_pieceinfo` where `exit_pieceinfo.requested_at` is not null and (`exit_pieceinfo.failed_at is not null and exit_pieceinfo.failed_status_code <> NOT_FOUND`) or `exit_pieceinfo.completed_at` is null
+        - If records are found, it should create new order limits for the found records, and return for reprocessing
+        - Else it returns a list of order limits for pieces to be moved to new nodes from  `exit_pieceinfo` ordered by `durability_ratio`. durability_ratio == num pieces / optimal number of pieces. lower values take presedence.
+  - ProcessPutOrders is called by the storagenode to send processed order limits. ProcessPutOrders verifies the hashes of successful orders limits, updates `exit_pieceinfo.completed_at`, and updates metainfo with the new piece location. Failures update `exit_pieceinfo.failed_at`  `exit_pieceinfo.failed_status_code`, for reprocessing. 
+    - If the failed status code is NOT_FOUND, the exiting storage node no longer stores that piece. In this case the node/piece should be removed from the pointer so that a repair is triggered.
+  - ``` 
+	service GracefulExit {
+		// Called by the storagenode to initiate an exit request
+		rpc InitiateExit(InitiateExitRequest) returns (InitiateExitResponse) {}
```

Comment:
Having initiate exit and get put orders separate would be more complicated to write. It'll probably easier to have single streaming rpc.